### ISSUE TITLE
Replicating Grave's Generative Wikipedia Model

### ISSUE THREAD

**Issue by RaffEdwardBAH** [fchollet/keras#1032]
I'm attempting to replicate the results in [Generating Sequences With Recurrent Neural Networks, by Alex Graves](http://arxiv.org/pdf/1308.0850v5.pdf) for a generative model on Wikipedia. The generative example [here ](https://github.com/fchollet/keras/blob/master/examples/lstm_text_generation.py) is built by eating a sequence of _n_ characters and then attempting to predict the _n+1_th character. 

However, the paper by Graves (as I understand, please correct me if wrong) eats characters (with a zero vector as the first input) and outputs the same characters being eaten. ie: "*Hello World" would be the input characters and "Hello World" would be the outputs, where "*" would indicate the zero vector. Another note in the paper is that, to learn the long term structure, they eat 10k characters (100 byte sequences with internal state only reset every 100 sequences, and sequences are kept in order)  but only push the gradient back to 100  steps. 

So I started to set up a similar model as 

``` python
model = Sequential()
model.add(LSTM(2048, return_sequences=True, input_shape=(max_len, tokens), truncate_gradient=100))
model.add(Dropout(0.5))
model.add(LSTM(2048, return_sequences=True, truncate_gradient=100))
model.add(Dropout(0.5))
model.add(LSTM(2048, return_sequences=False, truncate_gradient=100))
model.add(Dropout(0.5))
model.add(Dense(tokens))
model.add(Activation('softmax'))

model.compile(loss='categorical_crossentropy', optimizer=Adam(clipnorm=1))
```

However I'm not sure this is correct, and I'm not sure what shape the target variable `y` should be. I think the training matrix would be shaped as `X = np.zeros((train_sequences, max_len+1, tokens), dtype=np.bool)`, where the `+1` would be for the zero vector start. 

I have all characters read into a string and my initialization looks like 

``` python
row= 0
for i in range(0, len(train_bytes), max_len):
    for t, byte in enumerate(train_bytes[i:i+max_len]):
        X[row, t+1, ord(byte)] = 1
    row += 1
```

Can anyone help me finish this / set me in the right direction? I'm not seeing how to indicate to keras that if I made `y` 3 dimensional how the targets line up with the inputs. ie: the fist "l" should be predicted after receiving "*He" and the gradient pushed back from there, and then the second "l" after "*Hel", and so on. 

I saw issue #395 , but it is set up somewhat different - and the issue quite cluttered at this point.

**Comment by fchollet** [comment]
> However I'm not sure this is correct,

As a model that maps a sequence of n characters to the (n+1)th character, it looks correct. However! Do not specify `input_shape=(max_len, tokens)` in the second layer. Only do that in the first layer.

> I'm not sure what shape the target variable y should be.

It should be the (n+1)th character, encoded as a binary vector of k dimensions (k = size of your vocabulary). Take a look at the LSTM text generation example we have.

**Comment by RaffEdwardBAH** [comment]
> Do not specify input_shape=(max_len, tokens) in the second layer. Only do that in the first layer.

Oops! Copy-pasta error. 

>As a model that maps a sequence of n characters to the (n+1)th character, it looks correct.   

Right, I understand that - but I want the training procedure to train on every character, since my sequence is now 100k long. As I understand the Graves paper, if I wanted to train on say "Hello World!" and only do BPTT for 2 steps, I want to train on the sequences:

* Input "\*" and output "H"
* Input "\*H" and output "e"
* Input "\*He" and output "l"
* Input "\*Hel" and output "l"
* Input "\*Hell" and output "o"
* Input "\*Hello" and output " "
* ...
* Input "\*Hello Worl" and output "d"
* Input "\*Hello World" and output "!"

I could produce several input output pairs this way, and use the same style code in the LSTM example already in Keras, but this would blow up the number of training points. This is because its learning a sequence mapped to a single output. What I want to replicate in the paper, as I understand it, is to train a sequence mapped to the same sequence (but lagging a character behind). The discussion in #395 is similar, but it appears they are trying to map a sequence to a dense representation, and then decode the dense representation back to a new sequence. 

Am I making sense?

**Comment by RaffEdwardBAH** [comment]
Ok, I've been playing around and I think I've got a toy example of what I want working. 

```python
txt = "Hello World!"

model = Sequential()
model.add(LSTM(64, return_sequences=True, input_shape=(len(txt)+1, tokens), truncate_gradient=-1))
model.add(TimeDistributedDense(tokens))
model.add(Activation('softmax'))

model.compile(loss='categorical_crossentropy', optimizer=Adam(clipnorm=1))

X = np.zeros((1, len(txt)+1, tokens), dtype=np.bool)
y = np.zeros((1, len(txt), tokens), dtype=np.bool)

for t, byte in enumerate(txt):
    X[0, t+1, ord(byte)] = 1
    y[0, t, ord(byte)] = 1

model.fit(X, y, batch_size=1, nb_epoch=3000)
```

Am i correct that, in this code, if I set `truncate_gradient=2`, that the gradient will be truncated with respect to its occurrence in the sequence? 

Ie: the "o" character will be truncated to the two "l"s, the " " would be truncated to "lo". 

If I sample from this

```python
for i in range(len(txt)):
    print(model.predict(X[:, 0:i+1, :], verbose=0).shape)
    print(chr(sample(model.predict(X[:, 0:i+1, :], verbose=0)[0][i], 0.7)))
```

I get the expect output, and predict returns a prediction for all the previous steps as well. Is there a way to predict on the next *unseen* token once I've trained a model?

**Comment by lemuriandezapada** [comment]
I found it a lot better (and faster) to let the last layer return_sequences = True and set the input sequence as data[:,:-1,:] and the output data as data[:,1:,:] and not truncate the gradient at all.

**Comment by RaffEdwardBAH** [comment]
@lemuriandezapada could you provide a code example? I'm not sure I understand where `data` is coming from. Are you saying you do the 100 sequences with 100 characters each, instead of doing a 10k sequence with truncation? 

Also, do you mean faster in computation or convergence (or both)? Thanks for helping!

**Comment by lemuriandezapada** [comment]
My code is ugly as fuck, and cleaning it to post it would take too long :))
What I do is I chop the one hot sequences in say 100 characters each randomly from the text corpus.
data will be of size (train_sequences, seq_len, vocab_size) and you do 1hot encoding on the vocab size. So you can just set return_sequences=True in your model, and then call model.fit(data[:,:-1,:],data[:,1:,:])
Basically setting as input all the first n-1 [0:n-1] chars and target output the last n-1[1:last] chars. Instead of the usual way of using [0:n-1] input and [n] output.

It's faster on convergence (you start seeing results after fewer iterations), because it propagates back errors from more than 1 char. The other way still computes the reservoirs at every step and propagates the error through it from the last step but ignores all the other potential chars. Computing time per iteration itself is pretty much the same.

**Comment by RaffEdwardBAH** [comment]
@lemuriandezapada sorry for the late reply and thank you for contributing, but I'm not sure I understand. If you’re training on only 100 character sequences, and your not somehow copying hidden state to the next "sequence", how would the model get a chance to learn dependencies that are longer (1000 time-steps back or more). 

>Basically setting as input all the first n-1 [0:n-1] chars and target output the last n-1[1:last] chars. Instead of the usual way of using [0:n-1] input and [n] output.

This is part of what was specified in the graves paper that I was trying to replicate, as the keras example only does from one output. I think I have that part working now (as in my toy example above)

**Comment by lemuriandezapada** [comment]
It wouldn't. :)
For 1000+ timesteps you'd chop them up in 1000+ character sequences.
Keras's lack of RTRL-like implementation makes having to chop your sequences in any way mandatory and there's no workaround that I can think of. You'd need to try to propagate your errors all the way far into the past somehow.
Or you can try implementing something fancy like this:
http://arxiv.org/pdf/1402.3511v1.pdf
Though I've no idea how to do those timestep skips in keras/theano.

**Comment by RaffEdwardBAH** [comment]
@lemuriandezapada Ah, the main goal here is to replicate Grave's results with learning dependencies over very long ranges. From what I can find, his approach of doing sequential batches and copying hidden state from one to the next is the standard approach, and should be much better from a performance / GPU memory use perspective. I don't think the CW-RNN would help me train on longer sequences since the GPU memory is my bottle neck here. 

Thank you for your help though!
